src/Device/Driver/FANET/Parser.cpp

```diff
@@ -0,0 +1,304 @@
+  A detailed list of copyright holders can be found in the file "AUTHORS".
+
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU General Public License
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+#include "Device.hpp"
+#include "NMEA/Info.hpp"
+#include "FANET/FanetAddress.hpp"
+
+bool
+FanetDevice::AddStationToList(StationsList &stations_list, double clock,
+                              const FanetStation station) {
+  FanetStation *fanet_slot = stations_list.FindStations(station.address);
+
+  if (fanet_slot == nullptr) {
+    fanet_slot = stations_list.AllocateStation();
+    if (fanet_slot == nullptr)
+      // no more slots available
+      return false;
+
+    fanet_slot->Clear();
+    fanet_slot->address = station.address;
+    fanet_slot->location = station.location;
+    fanet_slot->location_available = station.location_available;
+  }
+
+  fanet_slot->valid.Update(clock);
+  fanet_slot->Update(station);
+
+  return true;
+}
+
+bool
+FanetDevice::IntegrityCheck(char *cmd_str) {
+  if (strlen(cmd_str) == 0)
+    return false;
+
+  for(char *ptr = cmd_str; *ptr != '\0'; ptr++)
+  {
+    if((*ptr >= '0' && *ptr <= '9') ||
+       (*ptr >= 'A' && *ptr <= 'F') ||
+       (*ptr >= 'a' && *ptr <= 'f') ||
+       (*ptr == ',')) {
+      continue;
+    }
+    return false;
+  }
+  return true;
+}
+
+char *
+FanetDevice::RemoveExtraChars(char *cmd_str) {
+  char *ptr = strchr(cmd_str, '\r');
+  if(ptr == nullptr)
+    ptr = strchr(cmd_str, '\n');
+  if(ptr != nullptr)
+    *ptr = '\0';
+  while(*cmd_str == ' ')
+    cmd_str++;
+  return cmd_str;
+}
+
+GeoPoint
+FanetDevice::Payload2GeoPoint(const uint8_t *buf) {
+  if(buf == nullptr)
+    return GeoPoint::Invalid();
+
+  // Integer values
+  unsigned int lati = buf[2] << 16 | buf[1] << 8 | buf[0];
+  if(lati & 0x00800000)
+    lati |= 0xFF000000;
+  unsigned int loni = buf[5] << 16 | buf[4] << 8 | buf[3];
+  if(loni & 0x00800000)
+    loni |= 0xFF000000;
+
+  return GeoPoint(Angle::Degrees((float)loni / 46603.0f), Angle::Degrees((float)lati / 93206.0f));
+}
+
+float
+FanetDevice::Payload2Float(uint8_t byte, float scale) {
+  auto value = (float)(byte & 0x7F);
+  if(byte & 1 << 7)
+    return (value * scale);
+  else
+    return value;
+}
+
+bool
+FanetDevice::DecodeServicePayload(uint8_t *payload, int16_t payload_length, FanetStation *station) {
+  if(payload == nullptr || payload_length == 0)
```

Comment:
Why is it legal to call `DecodeServicePayload(nullptr)`? This should be considered a bug, and thus a runtime check is the wrong way to check - do `assert(payload != nullptr)` instead.